File: app/src/main/java/com/example/calculator/layouts/Calculator.java
package com.example.calculator.layouts;

import android.annotation.SuppressLint;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;

import com.example.calculator.databinding.ActivityMainBinding;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textview.MaterialTextView;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

/**
 * TODO
 * Create exponent
 * Comment more
 * Test Parenthesis
 * Fix equals onClick
 */

public class Calculator {

    private static MaterialTextView display;
    private static ScriptEngine engine;

    private static String currentNumber = "";
    private static String displayText = "";
    private static String equation = "";

    /**
     * Initialize the Calculator
     * @param binding ActivityMainBinding
     */
    public static void init(@NonNull ActivityMainBinding binding) {

        // Store binding from activity
        display = binding.tvMain;

        // Get rhino JavaScript engine
        engine = new ScriptEngineManager().getEngineByName("rhino");

        // Initialize on click listeners for buttons
        initButtons(binding);

    }

    /**
     * Initialize button on click listeners
     * @param binding ActivityMainBinding
     */
    private static void initButtons(@NonNull ActivityMainBinding binding) {

        // Array of number buttons
        final MaterialButton[] numberButtons = new MaterialButton[] {

                binding.mbtnZero,
                binding.mbtnOne,
                binding.mbtnTwo,
                binding.mbtnThree,
                binding.mbtnFour,
                binding.mbtnFive,
                binding.mbtnSix,
                binding.mbtnSeven,
                binding.mbtnEight,
                binding.mbtnNine

        };

        // For each number button, set on click listener
        for (MaterialButton button : numberButtons)
            button.setOnClickListener(Calculator::onClickNumber);

        // Array of mathematical operators (basic)
        final MaterialButton[] operatorButtons = new MaterialButton[] {

                binding.mbtnPlus,
                binding.mbtnMinus,
                binding.mbtnMultiplication,
                binding.mbtnDivide

        };

        // For each mathematical operator, set on click listener
        for (MaterialButton button : operatorButtons)
            button.setOnClickListener(Calculator::onClickOperator);

        // Negate
        binding.mbtnNegate.setOnClickListener(Calculator::onClickNegate);

        // Decimal Point
        binding.mbtnPoint.setOnClickListener(Calculator::onClickPoint);

        // Delete
        binding.mbtnDelete.setOnClickListener(Calculator::onClickDelete);

        // Clear
        binding.mbtnClear.setOnClickListener(Calculator::onClickClear);

        // Equals
        binding.mbtnEquals.setOnClickListener(Calculator::onClickEquals);

        // Start Parenthesis
        binding.mbtnStartParen.setOnClickListener(Calculator::onClickParenthesis);

        // End Parenthesis
        binding.mbtnEndParen.setOnClickListener(Calculator::onClickParenthesis);

        // Exponent
        binding.mbtnExponent.setOnClickListener(Calculator::onClickExponent);

    }

    /**
     * Update the displayed text
     */
    private static void updateDisplay() {

        display.setText(String.format("%s %s", displayText, currentNumber));

    }

    /**
     * Logic for number buttons
     * @param v Button that is clicked
     */
    private static void onClickNumber(View v) {

        currentNumber = currentNumber + ((Button) v).getText().toString();

        updateDisplay();

    }

    /**
     * Logic for the negate button
     * @param v The negate button
     */
    private static void onClickNegate(View v) {

        if (!currentNumber.equals("")) {

            Double number = Double.parseDouble(currentNumber);

            number = number * -1;

            if (number % 1 != 0)
                currentNumber = String.valueOf(number);

            else currentNumber = String.valueOf(number.intValue());

            updateDisplay();

        }

    }

    /**
     * Logic for the decimal point button
     * @param v The decimal point button
     */
    private static void onClickPoint(View v) {

        if (!currentNumber.contains(".")) {

            currentNumber = currentNumber + ".";

            updateDisplay();

        }

    }

    /**
     * Logic for the delete button
     * @param v The delete button
     */
    private static void onClickDelete(View v) {

        if (!currentNumber.equals("")) {

            currentNumber = currentNumber.substring(0, currentNumber.length() - 1);

            updateDisplay();

        }

    }

    /**
     * Logic for the clear button
     * @param v The clear button
     */
    private static void onClickClear(View v) {

        currentNumber = "";
        displayText = "";
        equation = "";

        updateDisplay();

    }

    /**
     * Logic for basic operator buttons
     * @param v The operator button
     */
    private static  void onClickOperator(View v) {

        if (!currentNumber.equals("")) {

            if (displayText.equals(""))
                displayText = String.format("%s %s", currentNumber, ((Button) v).getText().toString());

            else if (equation.equals(""))
                displayText = String.format("%s %s %s", displayText, currentNumber, ((Button) v).getText().toString());

            else {

                displayText = String.format("%s %s %s", displayText, currentNumber, ((Button) v).getText().toString());
                equation = String.format("%s %s %s", equation, currentNumber, ((Button) v).getText().toString());

            }

            currentNumber = "";

            updateDisplay();

        } else if (!displayText.equals("")) {

            displayText = String.format("%s %s", displayText, ((Button) v).getText().toString());

            updateDisplay();

        }

    }

    /**
     * Logic for the equals button
     * @param v The equals button
     */
    @SuppressLint("SetTextI18n")
    private static void onClickEquals(View v) {

        Log.d("DEBUG_INFO", "Current Number: " + currentNumber);
        Log.d("DEBUG_INFO", "Display Text: " + displayText);
        Log.d("DEBUG_INFO", "Equation: " + equation);

        if (!currentNumber.equals("") || !displayText.equals(""))
            try {

                String result;

                if (equation.equals(""))
                    result = String.valueOf(engine.eval(displayText + currentNumber));

                else result = String.valueOf(engine.eval(equation + currentNumber));

                if ((Double.parseDouble(result) % 1 == 0))
                    result = String.valueOf(((Double) Double.parseDouble(result)).intValue());

                display.setText(String.format("%s %s = %s", displayText, currentNumber, result));

                displayText = "";
                currentNumber = result;

                if (result.equals("Infinity")) {

                    onClickClear(v);

                    display.setText("Cannot divide by zero");

                }

                Log.d("DEBUG_INFO", result);

            } catch (ScriptException e) {

                Log.wtf("onClickEquals", e);

                onClickClear(v);

                display.setText("Error");

            }

    }

    /**
     * Logic for the parenthesis buttons
     * @param v The parenthesis button
     */
    private static void onClickParenthesis(View v) {

        String button = ((Button) v).getText().toString();

        if (currentNumber.equals("")) {

            displayText = String.format("%s %s", displayText, button);

            if (!equation.equals(""))
                equation = String.format("%s %s", equation, button);

        } else {

            displayText = String.format("%s %s %s", displayText, currentNumber, button);

            if (!equation.equals(""))
                equation = String.format("%s %s %s", equation, currentNumber, button);

            currentNumber = "";

        }

        updateDisplay();

    }

    /**
     * Logic for the exponent button
     * @param v The exponent button
     */
    private static void onClickExponent(View v) {

        if (!currentNumber.equals("")) {

            if (equation.equals(""))
                equation = String.format("%s %s", displayText, String.format("Math.pow(%s,", currentNumber));

            else
                equation = String.format("%s %s", equation, String.format("Math.pow(%s,", currentNumber));

            displayText = String.format("%s %s", displayText, String.format("%s^(", currentNumber));
            currentNumber = "";

            updateDisplay();

        }

    }

}

File: kotlincalculator/src/main/java/com/example/kotlincalculator/layouts/Calculator.kt
package com.example.kotlincalculator.layouts

import android.util.Log
import android.view.View
import android.widget.Button
import com.example.kotlincalculator.databinding.ActivityMainBinding
import com.google.android.material.textview.MaterialTextView
import org.mozilla.javascript.Context
import java.lang.Exception


class Calculator(binding: ActivityMainBinding) {

    private val display: MaterialTextView = binding.tvMain
    private val context = Context.enter()

    private var currentNumber = ""
    private var displayText = ""
    private var equation = ""

    init {

        context.optimizationLevel = -1

        initButtons(binding)

    }

    /**
     * Initialize on click listeners for buttons
     * @param binding The main activity's binding
     */
    private fun initButtons(binding: ActivityMainBinding) {

        // Array of number buttons
        val numberButtons = arrayOf(

            binding.mbtnZero,
            binding.mbtnOne,
            binding.mbtnTwo,
            binding.mbtnThree,
            binding.mbtnFour,
            binding.mbtnFive,
            binding.mbtnSix,
            binding.mbtnSeven,
            binding.mbtnEight,
            binding.mbtnNine

        )

        // Array of mathematical operators (basic)
        val operatorButtons = arrayOf(

            binding.mbtnPlus,
            binding.mbtnMinus,
            binding.mbtnMultiplication,
            binding.mbtnDivide,
            binding.mbtnModulus

        )

        // For each number button, set on click listener
        for (button in numberButtons)
            button.setOnClickListener(this::onClickNumber)

        // For each operator button (basic), set on click listener
        for (button in operatorButtons)
            button.setOnClickListener(this::onClickOperator)

        // Negate
        binding.mbtnNegate.setOnClickListener(this::onClickNegate)

        // Decimal Point
        binding.mbtnPoint.setOnClickListener(this::onClinkPoint)

        // Delete
        binding.mbtnDelete.setOnClickListener(this::onClickDelete)

        // Clear
        binding.mbtnClear.setOnClickListener(this::onClickClear)

        // Start Parenthesis
        binding.mbtnStartParen.setOnClickListener(this::onClickParenthesis)

        // End Parenthesis
        binding.mbtnEndParen.setOnClickListener(this::onClickParenthesis)

        // Exponent
        binding.mbtnExponent.setOnClickListener(this::onClickExponent)

        // Equals
        binding.mbtnEquals.setOnClickListener(this::onClickEquals)

        // Square Root
        binding.mbtnSquareRoot.setOnClickListener(this::onClickSquareRoot)

    }

    /**
     * Evaluates a JavaScript script
     * @param script The script to execute
     * @return String or null
     */
    private fun evaluateScript(script: String): String? {

        try {

            // Create scope that the script wil run in
            val scope = context.initStandardObjects()

            // Run the script
            val result = context.evaluateString(scope, script, "script", 0, null)

            Log.d("DEBUG_INFO", result.toString())

            return result.toString()

        } catch(e: Exception) {

            Log.d("DEBUG_INFO", e.toString())

            return null

        }

    }

    /**
     * Update the displayed text
     */
    private fun updateDisplay() {

        display.text = String.format("%s %s", displayText, currentNumber)

    }

    /**
     * Logic for the number buttons
     * @param v The number button
     */
    private fun onClickNumber(v: View) {

        val button = (v as Button).text.toString()

        currentNumber = String.format("%s%s", currentNumber, button)

        updateDisplay()

    }


    /**
     * Logic for operator buttons (basic)
     * @param v The operator button
     */
    private fun onClickOperator(v: View) {

        val button = (v as Button).text.toString()

        if (currentNumber != "") {

            when {

                displayText == "" -> displayText = String.format("%s %s", currentNumber, button)
                equation == "" -> displayText = String.format("%s %s %s", displayText, currentNumber, button)

                else -> {

                    displayText = String.format("%s %s %s", displayText, currentNumber, button)
                    equation = String.format("%s %s %s", equation, currentNumber, button)

                }

            }

            currentNumber = ""

        } else if (displayText != "")
            displayText = String.format("%s %s", displayText, button)

        updateDisplay()

    }

    /**
     * Logic for the negate button
     * @param v The negate button
     */
    private fun onClickNegate(v: View) {

        if (currentNumber != "") {

            val number = currentNumber.toDouble() * -1

            currentNumber = if (number % 1.0 != 0.0) number.toString() else number.toInt().toString()

            updateDisplay()

        }

    }

    /**
     * Logic for the decimal point button
     * @param v The decimal point number
     */
    private fun onClinkPoint(v: View) {

        if (currentNumber != "" && !currentNumber.contains(".")) {

            currentNumber = String.format("%s.", currentNumber)

            updateDisplay()

        }

    }

    /**
     * Logic for the delete button
     * @param v The delete button
     */
    private fun onClickDelete(v: View) {

        if (currentNumber != "") {

            currentNumber = currentNumber.substring(0, currentNumber.lastIndex)

            updateDisplay()

        }

    }

    /**
     * Logic for the clear button
     * @param v The clear button
     */
    private fun onClickClear(v: View) {

        currentNumber = ""
        displayText = ""
        equation = ""

        updateDisplay()

    }

    /**
     * Logic for the parenthesis buttons
     * @param v The parenthesis button
     */
    private fun onClickParenthesis(v: View) {

        val button = (v as Button).text.toString()

        if (currentNumber == "") {

            displayText = String.format("%s %s", displayText, button)

            if (equation != "")
                equation = String.format("%s %s", equation, button)

        } else {

            displayText = String.format("%s %s %s", displayText, currentNumber, button)

            if (equation != "")
                equation = String.format("%s %s %s", equation, currentNumber, button)

            currentNumber = ""

        }

        updateDisplay()

    }

    /**
     * Logic for the exponent button
     * @param v The exponent button
     */
    private fun onClickExponent(v: View) {

        if (currentNumber != "") {

            equation = if (equation == "")
                String.format("%s %s", displayText, String.format("Math.pow( %s,", currentNumber))
                else String.format("%s %s", equation, String.format("Math.pow( %s,", currentNumber))

            displayText = String.format("%s %s", displayText, String.format("%s^(", currentNumber))
            currentNumber = ""

            updateDisplay()

        }

    }

    /**
     * Logic for the square root button
     * @param v The square root button
     */
    private fun onClickSquareRoot(v: View) {

        if (currentNumber == "") {

            equation = if (equation == "")
                String.format("%s Math.sqrt( ", displayText)
                else String.format("%s Math.sqrt( ", equation)

            displayText = String.format("%s √(", displayText)

            updateDisplay()

        }

    }

    /**
     * Logic for the equals button
     * @param v The equals button
     */
    private fun onClickEquals(v: View) {

        Log.d("DEBUG_INFO", "Current Number: $currentNumber")
        Log.d("DEBUG_INFO", "Display Text: $displayText")
        Log.d("DEBUG_INFO", "Equation: $equation")

        if (currentNumber != "" || displayText != "") {

            var result = if (equation == "")
                evaluateScript(String.format("%s %s", displayText, currentNumber))
                else evaluateScript(String.format("%s %s", equation, currentNumber))

            if (result != null) {

                if (result.toDouble() % 1.0 == 0.0)
                    result = result.toDouble().toInt().toString()

                display.text = String.format("%s %s = %s", displayText, currentNumber, result)

                displayText = ""
                equation = ""
                currentNumber = result

                if (result == "Infinity") {

                    onClickClear(v)

                    display.text = "Cannot divide by zero"

                }

            } else {

                onClickClear(v)

                display.text = "Invalid equation"

            }

        }

    }

}